<a href="https://colab.research.google.com/github/DavidVollendroff/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [0]:
### YOUR CODE STARTS HERE
import pandas as pd
from scipy.stats import ttest_ind

column_headers = ['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa']
data_url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data'

df = pd.read_csv(data_url, header=None, na_values="?", names=column_headers)
df.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n          NaN
2    democrat                 NaN             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [0]:
df = df.replace({'y':1, 'n': 0})
df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
rep_df = df[df['party']=='republican']
dem_df = df[df['party']=='democrat']
rep_df.drop('party', axis=1)
#rep_df['handicapped-infants'].mean()

handicapped-infants  water-project  budget  ...  crime  duty-free  south-africa
0                    0.0            1.0     0.0  ...    1.0        0.0           1.0
1                    0.0            1.0     0.0  ...    1.0        0.0           NaN
7                    0.0            1.0     0.0  ...    1.0        NaN           1.0
8                    0.0            1.0     0.0  ...    1.0        0.0           1.0
10                   0.0            1.0     0.0  ...    1.0        0.0           0.0
..                   ...            ...     ...  ...    ...        ...           ...
427                  0.0            0.0     0.0  ...    1.0        0.0           1.0
430                  0.0            0.0     1.0  ...    1.0        0.0           1.0
432                  0.0            NaN     0.0  ...    1.0        0.0           1.0
433                  0.0            0.0     0.0  ...    1.0        0.0           1.0
434                  0.0            1.0     0.0  ...    1.0        NaN           0.0

[168 rows x 16 columns]

In [0]:
def df_ttest(dataframe):
  rep_df = dataframe[dataframe['party']=='republican'].drop('party', axis=1)
  dem_df = dataframe[dataframe['party']=='democrat'].drop('party', axis=1)
  p_values = {}
  for column in rep_df:
    _, p_val = ttest_ind(rep_df[column], dem_df[column], nan_policy='omit')
    p_values[column] = p_val
  return p_values
df_ttest(df)

{'aid-to-contras': 2.82471841372357e-54,
 'anti-satellite-ban': 8.521033017443867e-31,
 'budget': 2.0703402795404463e-77,
 'crime': 9.952342705606092e-47,
 'duty-free': 5.997697174347365e-32,
 'education': 1.8834203990450192e-64,
 'el-salvador-aid': 5.600520111729011e-68,
 'handicapped-infants': 1.613440327937243e-18,
 'immigration': 0.08330248490425066,
 'mx-missile': 5.03079265310811e-47,
 'physician-fee-freeze': 1.994262314074344e-177,
 'religious-groups': 2.3936722520597287e-20,
 'right-to-sue': 1.2278581709672758e-34,
 'south-africa': 3.652674361672226e-11,
 'synfuels': 1.5759322301054064e-15,
 'water-project': 0.9291556823993485}

In [0]:
def support_level(dataframe):
  rep_df = dataframe[dataframe['party']=='republican'].drop('party', axis=1)
  dem_df = dataframe[dataframe['party']=='democrat'].drop('party', axis=1)
  rep_support_values = {}
  dem_support_values = {}
  for column in rep_df:
    rep_support_values[column] = rep_df[column].mean()
    dem_support_values[column] = dem_df[column].mean()
  return rep_support_values, dem_support_values

support_level(df)

Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01


Democrats showed more support for aid to contras than Republicans did, with a corresponding p-score of 2.82e-54. This means it is extremely unlikely that the votes casts by either party could have been drawn from a distribution with the same mean value of support.

Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01


Republicans showed more support for the bill having to do with crime than Democrats did, with a corresponding p-score of 9.95e-47. This means it is extremely unlikely that the votes casts by either party could have been drawn from a distribution with the same mean value of support.

Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

One could say with 92.9% confidence that support for the "water project" did not have a link to party affiliation. Similar levels of support from each party lead to a very high p-score of p = 0.93, far in excess of the p > 0.1 standard we sought.